libvast/src/format/arrow.cpp

```diff
@@ -74,4 +75,109 @@ bool writer::layout(const std::shared_ptr<::arrow::Schema>& schema) {
+  return pos_;
+}
+
+::arrow::Result<int64_t>
+arrow_istream_wrapper::Read(int64_t nbytes, void* out) {
+  input_->read(reinterpret_cast<char*>(out), nbytes);
+  if (input_) {
+    pos_ += nbytes;
+    return nbytes;
+  }
+  return 0;
+}
+
+::arrow::Result<std::shared_ptr<::arrow::Buffer>>
+arrow_istream_wrapper::Read(int64_t nbytes) {
+  ARROW_ASSIGN_OR_RAISE(auto buffer, ::arrow::AllocateResizableBuffer(nbytes));
+  ARROW_ASSIGN_OR_RAISE(int64_t bytes_read,
+                        Read(nbytes, buffer->mutable_data()));
+  ARROW_RETURN_NOT_OK(buffer->Resize(bytes_read, false));
+  buffer->ZeroPadding();
+  return std::move(buffer);
+}
+
+reader::reader(const caf::settings& options, std::unique_ptr<std::istream> in)
+  : vast::format::reader(options),
+    input_(std::make_unique<arrow_istream_wrapper>(std::move(in))) {
+}
+
+caf::error
+reader::read_impl(size_t max_events, size_t max_slice_size, consumer& f) {
+  VAST_TRACE_SCOPE("{} {}", VAST_ARG(max_events), VAST_ARG(max_slice_size));
+  VAST_ASSERT(max_events > 0);
+  // TODO: we currently ignore `max_slize_size` because we're just passing
+  // through existing table slices / record batches from the producer system.
+  VAST_ASSERT(max_slice_size > 0);
+  auto reader = ::arrow::ipc::RecordBatchStreamReader::Open(input_.get());
+  if (!reader.ok()) {
+    return caf::make_error(ec::logic_error, fmt::format("error creating stream "
+                                                        "reader: '{}'",
+                                                        reader.status()));
+  }
+  size_t produced = 0;
+  while (produced < max_events) {
+    if (const auto batch = (*reader)->ReadNext(); batch.ok()) {
+      // When the schema changes and a new IPC message begins, we see one
+      // record batch with status `OK` and `nullptr` as data and re-initialize
+      // the reader.
+      if (batch->batch != nullptr) {
+        const auto slice = table_slice{batch->batch};
+        f(slice);
+        produced += slice.rows();
+      } else {
+        reader = ::arrow::ipc::RecordBatchStreamReader::Open(input_.get());
+        if (!reader.ok()) {
+          return caf::make_error(
+            ec::logic_error,
+            fmt::format("error creating stream reader: '{}'", reader.status()));
+        }
+      }
+    } else {
+      // Reading the next record batch yields an error if the input stream ends.
+      // We check if it's actually just eof or an actual error.
+      if (input_->closed()) {
+        return caf::make_error(ec::end_of_input, "input exhausted");
+      }
+      return caf::make_error(
+        ec::format_error,
+        fmt::format("failed to read next record batch '{}'", batch.status()));
+    }
+  }
+  return caf::none;
+}
+
+void reader::reset(std::unique_ptr<std::istream> in) {
+  input_ = std::make_unique<arrow_istream_wrapper>(std::move(in));
+}
+
+caf::error reader::module(vast::module x) {
+  module_ = x;
+  return caf::none;
+}
```

Comment:
We don't make use of this at all because we always use the schema attached to the incoming batches. Should this return an error instead, similar to the NetFlow/IPFIX reader?

```c++
caf::error reader::module([[maybe_unused]] class module module) {
  // The VAST types are automatically generated and cannot be changed.
  return caf::make_error(ec::no_error,
                         "schema cannot be changed as it is generated "
                         "dynamically");
}
```

Reply:
Thanks, I didn't see that! I was wondering, but couldn't escape the API design.

I'll add a similar error, even though `ec::no_error` doesn't strike me as very descriptive here.

Reply:
With this change, `vast spawn source arrow` now prints:

```
[14:14:34.079] arrow-reader source failed to set schema: !! logic_error: schema is derived from the Arrow input and can't be changed
```

I don't understand the differences between `import` and `spawn source` enough to understand why it doesn't happen for `vast import`. I wonder if a debug-level log message may be the better approach to prevent this error from popping up. FWIW, the data _is_ actually imported with `vast spawn source arrow`.

Reply:
I think you actually have to use `ec::no_error` here. That is the only alternative that makes the `caf::error` false-y.

Reply:
Switched to `ec::no_error` and the log message disappears.